File: api-certvalidation/csharp/src/find/ItemFinderIteration.cs
using System.Collections.Generic;
using tr.gov.tubitak.uekae.esya.api.common;
using tr.gov.tubitak.uekae.esya.api.common.util;

namespace tr.gov.tubitak.uekae.esya.api.certificate.validation.find
{
    /**
     * @author ayetgin
     */
    public abstract class ItemFinderIteration<T> where T : class
    {
        protected List<ItemSource<T>> mSources = new List<ItemSource<T>>();

        protected ItemSource<T> mCurrentSource;
        protected T mCurrentItem;

        public virtual bool nextIteration(ValidationSystem aValidationSystem)
        {
            T pItem = null;
            //T pItem = default(T);

            if (mCurrentSource != null)
            {
                while ((pItem = mCurrentSource.nextItem()) != null)
                {
                    mCurrentItem = pItem;
                    return true;
                }
            }

            while (_nextSource(aValidationSystem))
            {
                while ((pItem = mCurrentSource.nextItem()) != null)
                {
                    mCurrentItem = pItem;
                    return true;
                }
            }
            return false;
        }

        public T getCurrentItem()
        {
            if (mCurrentItem != null)
                return mCurrentItem;
            else throw new ESYARuntimeException("Iteration out of bounds");
        }

        public void addItemSource(ItemSource<T> aItemSource)
        {
            mSources.Add(aItemSource);
        }

        protected abstract void _initSources(ValidationSystem aValidationSystem);

        protected abstract bool _nextSource(ValidationSystem aValidationSystem);
    }
}

File: api-certvalidation/src/tr/gov/tubitak/uekae/esya/api/certificate/validation/find/ItemFinderIteration.java
package tr.gov.tubitak.uekae.esya.api.certificate.validation.find;

import tr.gov.tubitak.uekae.esya.api.certificate.validation.ValidationSystem;
import tr.gov.tubitak.uekae.esya.api.common.ESYAException;
import tr.gov.tubitak.uekae.esya.api.common.ESYARuntimeException;
import tr.gov.tubitak.uekae.esya.api.common.util.ItemSource;

import java.util.ArrayList;
import java.util.List;

/**
 * @author ayetgin
 */
public abstract class ItemFinderIteration<T>
{
    protected List<ItemSource<T>> mSources = new ArrayList<ItemSource<T>>();

    protected ItemSource<T> mCurrentSource;
    protected T mCurrentItem;

    public boolean nextIteration(ValidationSystem aValidationSystem) throws ESYAException
    {
        T pItem = null;

        if (mCurrentSource != null) {
            while ((pItem = mCurrentSource.nextItem()) != null) {
                mCurrentItem = pItem;
                return true;
            }
        }

        while (_nextSource(aValidationSystem)) {
            while ((pItem = mCurrentSource.nextItem())!=null) {
                mCurrentItem = pItem;
                return true;
            }
        }
        return false;
    }

    public T getCurrentItem()
    {
        if (mCurrentItem != null)
            return mCurrentItem;
        else throw new ESYARuntimeException("Iteration out of bounds");
    }

    public void addItemSource(ItemSource<T> aItemSource){
        mSources.add(aItemSource);
    }

    protected abstract void _initSources(ValidationSystem aValidationSystem);

    protected abstract boolean _nextSource(ValidationSystem aValidationSystem);

}
